### ISSUE TITLE
No documentation on "Gtk Composite Templates" has been added yet

### ISSUE THREAD

**Issue by sharkwouter** [sebp/PyGObject-Tutorial#149]
Since a about 8 months ago, PyGObject has supported GTK composite templates. This works similarly to the features in Vala. It allows users to use ui/glade files as templates for their own classes. See the following issue on the gnome gitlab for more information: https://gitlab.gnome.org/GNOME/pygobject/issues/52

I have not been able to find any resources for this feature and I haven't seen any projects which use it yet, but here is an example:
```python
import gi
import sys

gi.require_version('Gtk', '3.0')

from gi.repository import Gio
from gi.repository import Gtk

@Gtk.Template.from_file("hello-window.ui")
class Window(Gtk.Window):
    __gtype_name__ = "HelloWindow"

    headerbar = Gtk.Template.Child()
    label = Gtk.Template.Child()

def create_window(app):
    win = Window(application=app, default_width=600, default_height=300)
    return win

def on_activate(app):
    win = app.get_active_window()
    if win is None:
        win = create_window(app)
    win.present()

if __name__ == '__main__':
    app = Gtk.Application.new('com.example.Hello', Gio.ApplicationFlags.FLAGS_NONE)
    app.connect('activate', on_activate)
    sys.exit(app.run())
```
The ui file:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<interface>
  <template class="HelloWindow" parent="GtkWindow">
    <property name="default_width">600</property>
    <property name="default_height">300</property>
    <child type="titlebar">
      <object class="GtkHeaderBar" id="headerbar">
        <property name="visible">true</property>
        <property name="show-close-button">true</property>
        <property name="title">Hello, World!</property>
      </object>
    </child>
    <child>
      <object class="GtkLabel" id="label">
        <property name="label">Hello, World!</property>
        <property name="visible">true</property>
        <attributes>
          <attribute name="weight" value="bold"/>
          <attribute name="scale" value="2"/>
        </attributes>
      </object>
    </child>
  </template>
</interface>
```
These examples are an adaption on the example found in the following repo: https://gitlab.gnome.org/chergert/hello

I have not tested to see if this feature works with older GTK versions. It works with the latest stable release. I've tested in on Ubuntu 18.10, Arch and MacOS Mojave.

It might be a good idea to add this feature to the Glade documentation. I may be able to put some work into this later, if this is seen as useful.

**Comment by sebp** [comment]
Sorry for the late reply, if you could create a pull request with some text and your example, I'd be happy to merge it.

**Comment by ismailarilik** [comment]
Thanks a lot. You may think to add a Callback usage also.

**Comment by rscmbbng** [comment]
Hi all, I'm also looking for examples of Composite Templates. I have found the following example in the Librem 5 docs: https://developer.puri.sm/Librem5/Apps/Examples/Audio/Play_Sounds/Source_Code.html

It both uses the `Gtk.Template` and `Gtk.Template.Callback`. I'm still trying to wrap my head around it but it seems like a good start for writing a tutorial.

**Comment by sharkwouter** [comment]
@rscmbbng since I'm using this feature for one of my projects (a GOG client called Minigalaxy) and it has some contributors I made some very short instructions on how to work with it. Maybe it is of some use to you. It can be found here: https://github.com/sharkwouter/minigalaxy/wiki/Developer-information#glade-usage

It is very short, though. It doesn't really replace proper documentation.

**Comment by rscmbbng** [comment]
Hi @sharkwouter that repository is super helpful! Thanks for sharing.

**Comment by cjmayo** [comment]
Upstream merge request for an attempt at Gtk.Template documentation:

https://gitlab.gnome.org/GNOME/pygobject/-/merge_requests/162

**Comment by sebp** [comment]
@cjmayo I'm not maintaining the upstream pygobject documentation. If you want to integrate it into this tutorial, I'd be happy to merge it if you create a pull request on GitHub.

**Comment by cjmayo** [comment]
Thanks, I was thinking it should have some upstream documentation first - to make sure this really is a stable API.

**Comment by sebp** [comment]
This is now fixed by merging #195
